Add IDManager and named tempTimers/event handlers from script

demonnic opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions

Adds named tempTimers and event handlers to the Lua API. If you register to the same name more than once, it handles the details of killing the previous timer or handler and reregistering rather than allowing duplication. Works with oneShot handlers and timers, as it's still using tempTimer() and registerAnonymouseEventHandler() under the hood.

Also adds an IDManager class which is what actually handles all of this, and allows script authors to get their own IDManager if they want using getNewIDManager().  ~~I still need to write a wiki page around the manager.~~ The manager docs are sitting at https://wiki.mudlet.org/w/Area_51/IDManager until release, then they'll get a `Manual:` page

Also adds the following functions for events. ~~These all use a default IDManager which is created during profile load, so do not touch any other IDManager which may have been created, they're fully independent.~~ This creates an IDManager for each 'user' and manages each user separately to enforce namespacing and prevent collisions. In all cases if it is the first time a userName has been used, a new IDManager is create for them so there is no need to worry about registrating.
  * registerNamedEventHandler(userName, handlerName, eventName, funcreference or string, oneShot)
    * registers a named event handler. Returns true if it worked, otherwise printErrors
  * stopNamedEventHandler(userName, handlerName)
    * kills the named handler but holds on to the information for resuming
  * resumeNamedEventHandler(userName, handlerName)
    * re-registers the event handler with the stored information
  * deleteNamedEventHandler(userName, handlerName)
    * kills the named handler and removes it from the manager for userName entirely, as though it never happened
  * stopAllNamedEventHandlers(userName)
    * stops all the named event handlers for userName
  * deleteAllNamedEventHandlers(userName)
    * I too like to live dangerously: deletes all named event handlers for userName
  
There is a full complement of the same for timers as registerNamedTimer(), resumeNamedTimer() etc.

#### Motivation for adding to Mudlet

One of the most common mistakes for new Mudlet scripters is not catching and handling the return ID from tempTimer or registerAnonymousEventHandler. They end up with a bunch of duplicates they can't stop and have to restart their profile, and get help from us, and then get told there's all this scaffolding code they have to do for each tempTimer or eventHandler they want.

So I thought
![image](https://user-images.githubusercontent.com/3660/137595424-8ff779b6-f62b-4c0d-a31b-89a494cf9031.png)

When someone asked about named event handlers and such again last night, and wrote the IDManager and the functions to use a default one.

#### Other info (issues closed, discussion etc)

I wrote tests for the entire event pathway, but the tests for the timers are pending me figuring out async busted tests and I wanted to go ahead and get the PR opened. I also added a TODO for it, will open the issue to track it with.

demonnic committed:
Add IDManager and API to either make one to control yourself or register things to a profile's default manager. Also, tests

demonnic committed:
Add issue link to TODO for passing danger checks

demonnic commented:
Staged IDManager docs at https://wiki.mudlet.org/w/Area_51/IDManager and the API calls to use the 'default' manager are staged in Area_51 as well.

vadi2 reviewed (commented):
✔ good idea from my side

demonnic committed:
Adjust to take a 'user' argument and create a manager per user to promote siloing. Add errors.

demonnic commented:
After some discussion with @vadi2 during the Mudlet Monday chat today we decided it's better to enforce some name spacing here, so I have adjusted it to create a new IDManager per 'user name' and require a userName argument. I have adjusted the tests accordingly.

I've also added errors for if either the timer/eventName or userName is missing or the wrong type. And tests for those. 

I will be adding tests for ensuring two usernames do not collide in a bit here as well.

demonnic committed:
Add test around separation of users' named event handlers

demonnic commented:
ok, test added.

demonnic commented:
area 51 docs for functions updated to reflect the new userName argument

demonnic committed:
Make handlerSpy a local

demonnic commented:
could I get someone from @Mudlet/lua-interface to look this over since I made the adjustment and give review? Would be nice to get in before feature freeze.

vadi2 reviewed (approved):
https://discord.com/channels/283581582550237184/427919962561052673/900774937025998848 I think is a good idea, but it's not a blocker. Let's get the testing team to approve this too

keneanung commented on src/mudlet-lua/lua/IDManager.lua:
```diff
@@ -0,0 +1,399 @@
+    timers = {}
+  }
+  setmetatable(mgr, self)
+  self.__index = self
```
This works, but using the metatable as the prototype is confusing to me... Am I the only one or can we split these responsibilities?

vadi2 replied:
I'm not strong on it either tbh

demonnic replied:
I'm not sure what you're suggesting? This is not only the pattern used all throughout Geyser, it's the one used in the PiL chapter on classes and objects.

demonnic replied:
link for reference https://www.lua.org/pil/16.1.html

keneanung replied:
That's how well I read PIL (or geyser source for that matter) :stuck_out_tongue: But for me, it's the Single Responsibility Principle (SRP): one thing to build the "link" (metastable) and one thing to have the prototype/class implementation. But since we already used that pattern, it's probably futile to argue against it :wink:

demonnic replied:
I'm not sure what that would look like in Lua. Are you suggesting to create the metatable, and then rather than include the new function in the metatable which sets the metatable, use an external function to create the new table and set its metatable? I did discover I could make a small optimization by exchanging `self.__index = self` in the new method with `IDManager.__index = IDManager` outside of the new method, which means it's only done once instead of once per instantiation.

keneanung replied:
My thinking was more along the lines of creating a metatable (can also be anonymous):
```lua
setmetatable(mgr, { __index = self })
```
(btw: PIL uses a different pattern in a different place: https://www.lua.org/pil/13.4.1.html)

demonnic replied:
This seems to me like it creates a bunch of extra table references to accomplish the same effect and is probably why the pattern we use now was referred to as an optimization in the later chapter on classes and inheritance.

I think let's leave it like it is for this PR and we can discuss if we want to change the style across the board and make a PR for that if we do decide to make that change.

keneanung reviewed (approved):
Looks good to me in general. One thing is violating the SRP for me, but if everybody else is ok with it. go for it.

vadi2 reviewed (approved):
Good with the changes & API then, most people prefer the two argument variant.
